bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/form/FormPostServiceImpl.java

```diff
@@ -0,0 +1,191 @@
+import org.apache.sling.servlets.post.JSONResponse;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.metatype.annotations.AttributeDefinition;
+import org.osgi.service.metatype.annotations.Designate;
+import org.osgi.service.metatype.annotations.ObjectClassDefinition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.adobe.cq.wcm.core.components.services.form.FormPostService;
+import com.day.cq.wcm.foundation.forms.FormsHelper;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+
+@Component(
+        service = {FormPostService.class}
+)
+@Designate(
+        ocd = FormPostServiceImpl.Config.class
+)
+public class FormPostServiceImpl implements FormPostService {
+
+    private static final int DEFAULT_CONNECTION_TIMEOUT = 6000;
+    private static final int DEFAULT_SOCKET_TIMEOUT = 6000;
+    private static final Logger LOG = LoggerFactory.getLogger(FormPostServiceImpl.class);
+    private static final Set<String> INTERNAL_PARAMETER = ImmutableSet.of(
+            ":formstart",
+            "_charset_",
+            ":redirect",
+            ":cq_csrf_token"
+    );
+
+    private static final String PN_FORM_ENDPOINT_URL = "formEndPointUrl";
+    private static final String CHARSET = "UTF-8";
+
+    private int connectionTimeout;
+    private int socketTimeout;
+
+    @Reference
+    private HttpClientBuilderFactory clientBuilderFactory;
+
+    @Override
+    public boolean sendFormData(SlingHttpServletRequest request, SlingHttpServletResponse response) {
+        try {
+            JSONObject jsonRequest = requestParamsToJSON(request);
+            return processFormApi(request, response, jsonRequest);
+        } catch (JSONException e) {
+            LOG.error(e.getMessage());
+        }
+        return false;
+    }
+
+    private boolean processFormApi(SlingHttpServletRequest request, SlingHttpServletResponse response, JSONObject jsonRequest) {
+        boolean processFormApiSuccess = false;
+        Resource formContainerResource = request.getResource();
+        ValueMap valueMap = formContainerResource.adaptTo(ValueMap.class);
+        if (valueMap != null) {
+            String formEndpoint = valueMap.get(PN_FORM_ENDPOINT_URL, String.class);
+            if (StringUtils.isNotEmpty(formEndpoint)) {
+                processFormApiSuccess = callFormApi(formEndpoint, jsonRequest);
+            }
+            sendRedirect(valueMap, request, response, processFormApiSuccess);
+        }
+        return processFormApiSuccess;
+    }
+
+    private void sendRedirect(ValueMap valueMap, SlingHttpServletRequest request, SlingHttpServletResponse response, boolean processFormApiSuccess) {
+        String redirect = valueMap.get("redirect", String.class);
```

Comment:
You have to check the redirect parameter. It should not redirect to other servers.

Reply:
I don't think that this is required as in L119 the redirect path is mapped and will return null if the given path can't be mapped to an internal resource.

Reply:
This only returns null if there is no mapping, but the default mapping will always return the same path. The javadoc explicitly says:
bq. Note, that technically the resourcePath need not refer to an existing resource. This method just applies the mappings and returns the resulting string. (https://github.com/apache/sling-org-apache-sling-resourceresolver/blob/master/src/main/java/org/apache/sling/resourceresolver/impl/ResourceResolverImpl.java#L420).

Reply:
Compare also with https://issues.apache.org/jira/browse/SLING-9850!

Reply:
added test for internal resource in 88bc244100983582704ffa7992390d717b9da37c